File: elixir/word-count/lib/word_count.ex
defmodule WordCount do
  @doc """
  Count the number of words in the sentence.
  Words are compared case-insensitively.
  """

  def count(sentence) do

    sentence
    |> String.downcase
    |> String.split(~r/[^[:alnum:]-]/u, trim: true)                                   # creates a list of words
                                                                                      # trims out empty list items
                                                                                      # regex:
                                                                                      # >> single space not alphanumeric or a hyphen
                                                                                      # >> unicode support solves German character problem
    |> Enum.reduce(%{}, fn word, acc -> Map.update(acc, word, 1, (&(&1 + 1))) end)

    # |> Enum.reduce(%{}, (&Map.update(&2, &1, 1, (&(&1 + 1))))) # <-- familed attempt to further simplify function

  end


end

File: javascript/word-count/word-count.js
export const countWords = (string) => { return dictBuild(string.trim(' ').split(/\b[^a-zA-Z]+\b/)); };

const dictBuild = (list, dict = []) => {
  let word = '';
  [word, ...list] = list;
  if(dict.indexOf(word) == -1) {dict[word] = 1; } 
  else { dict[word] += 1; }
  return list ? dictBuild(list, dict) : dict;
};